pkg/pipeline/hooks/drivers/bitbucketserver.go

```diff
@@ -1 +1,178 @@
 package drivers
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/rancher/rancher/pkg/pipeline/providers"
+	"github.com/rancher/rancher/pkg/pipeline/remote/bitbucketserver"
+	"github.com/rancher/rancher/pkg/pipeline/remote/model"
+	"github.com/rancher/rancher/pkg/pipeline/utils"
+	"github.com/rancher/rancher/pkg/ref"
+	"github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	BitbucketServerWebhookHeader  = "X-Request-Id"
+	bitbucketServerEventHeader    = "X-Event-Key"
+	bitbucketServerPushEvent      = "repo:refs_changed"
+	bitbucketServerPrCreatedEvent = "pr:opened"
+	bitbucketServerPrUpdatedEvent = "pr:modified"
+	bitbucketServerStateOpen      = "OPEN"
+)
+
+type BitbucketServerDriver struct {
+	PipelineLister             v3.PipelineLister
+	PipelineExecutions         v3.PipelineExecutionInterface
+	SourceCodeCredentials      v3.SourceCodeCredentialInterface
+	SourceCodeCredentialLister v3.SourceCodeCredentialLister
+}
+
+func (b BitbucketServerDriver) Execute(req *http.Request) (int, error) {
+	var signature string
+	if signature = req.Header.Get(githubSignatureHeader); len(signature) == 0 {
+		return http.StatusUnprocessableEntity, errors.New("webhook missing signature")
+	}
+	event := req.Header.Get(bitbucketServerEventHeader)
+	if event != bitbucketServerPushEvent && event != bitbucketServerPrCreatedEvent && event != bitbucketServerPrUpdatedEvent {
+		return http.StatusUnprocessableEntity, fmt.Errorf("not trigger for event:%s", event)
+	}
+
+	pipelineID := req.URL.Query().Get("pipelineId")
+	ns, name := ref.Parse(pipelineID)
+	pipeline, err := b.PipelineLister.Get(ns, name)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return http.StatusUnprocessableEntity, err
+	}
+	logrus.Infof("get body:%s", string(body))
+	if match := verifyBitbucketServerWebhookSignature([]byte(pipeline.Status.Token), signature, body); !match {
+		return http.StatusUnprocessableEntity, errors.New("invalid signature")
```

Comment:
do you think we can log of what exactly failed to validate? or the expected format (like < 71 in len, etc)